From 7b5bac193186beefa1724bc56d7b30e2f0a26575 Mon Sep 17 00:00:00 2001
From: Shayan Darayan <sdarayan@utexas.edu>
Date: Sun, 6 Feb 2022 20:04:44 +0000
Subject: [PATCH] Remove BYTE directives from kernel linker script to fix
 triple fault on boot

---
 kernel.ld | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/kernel.ld b/kernel.ld
index e24c860..4e12e14 100755
--- a/kernel.ld
+++ b/kernel.ld
@@ -26,16 +26,12 @@ SECTIONS
 		PROVIDE(__STAB_BEGIN__ = .);
 		*(.stab);
 		PROVIDE(__STAB_END__ = .);
-		BYTE(0)		/* Force the linker to allocate space
-				   for this section */
 	}
 
 	.stabstr : {
 		PROVIDE(__STABSTR_BEGIN__ = .);
 		*(.stabstr);
 		PROVIDE(__STABSTR_END__ = .);
-		BYTE(0)		/* Force the linker to allocate space
-				   for this section */
 	}
 
 	/* Adjust the address for the data segment to the next page */
-- 
2.25.1

